### PR TITLE
chore: upgrade to jest 24, update docs to reflect changes

### LINKED ISSUE #11267: Unit Testing tutorial yields failing test after following steps
## Description

I've attempted to follow the steps in the [Unit testing guide](https://www.gatsbyjs.org/docs/unit-testing/), as well as the [Testing components with GraphQL guide](https://www.gatsbyjs.org/docs/testing-components-with-graphql/), but the simple index.js test is still failing. I think the tutorial needs to be updated.

### Steps to reproduce

```
# install starter blog
npx gatsby new my-blog-starter https://github.com/gatsbyjs/gatsby-starter-blog

# start with the unit testing guide: https://www.gatsbyjs.org/docs/unit-testing/
# install dependencies while in the starter blog project
npm install --save-dev jest babel-jest react-test-renderer identity-obj-proxy babel-core@^7.0.0-bridge.0 @babel/core babel-preset-gatsby

# create jest.config.js as described
# create jest-preprocess.js as described
# create __mocks__/fileMock.js as described
# create loadershim.js as described
# create __mocks__/gatsby.js as described
# update test script in package.json to call jest
# create src/__tests__/index.js using provided code with location const
# try running npm run test, this fails (as expected) bc no GraphQL data being passed to component 
```
screen shot 2019-01-24 at 3 31 26 pm

```
# So now move to GraphQL guide: https://www.gatsbyjs.org/docs/testing-components-with-graphql/
# update src/__tests__/index.js with provided code. This code contains data const
# which was obtained from GraphiQL IDE
npm run test
```
Resulting Error: Invariant Violation: mockConstructor(...)
screen shot 2019-01-24 at 3 36 29 pm

### Expected result

After providing mocked data within the index.js test, I expected the test to pass and for there to be a snapshot in the snapshot dir as described:
screen shot 2019-01-24 at 3 46 44 pm

### Actual result

The test fails, see above screenshot for error message.

### Environment

 ```
 System:
    OS: macOS Sierra 10.12.6
    CPU: (4) x64 Intel(R) Core(TM) i7-4578U CPU @ 3.00GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 10.13.0 - /usr/local/bin/node
    Yarn: 1.13.0 - /usr/local/bin/yarn
    npm: 6.5.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 71.0.3578.98
    Firefox: 64.0
    Safari: 11.1.2
  npmPackages:
    gatsby: ^2.0.76 => 2.0.76
    gatsby-image: ^2.0.22 => 2.0.25
    gatsby-plugin-feed: ^2.0.8 => 2.0.11
    gatsby-plugin-google-analytics: ^2.0.5 => 2.0.8
    gatsby-plugin-manifest: ^2.0.5 => 2.0.12
    gatsby-plugin-offline: ^2.0.5 => 2.0.20
    gatsby-plugin-react-helmet: ^3.0.0 => 3.0.5
    gatsby-plugin-sharp: ^2.0.6 => 2.0.16
    gatsby-plugin-typography: ^2.2.0 => 2.2.3
    gatsby-remark-copy-linked-files: ^2.0.5 => 2.0.8
    gatsby-remark-images: ^2.0.4 => 2.0.6
    gatsby-remark-prismjs: ^3.0.0 => 3.1.4
    gatsby-remark-responsive-iframe: ^2.0.5 => 2.0.8
    gatsby-remark-smartypants: ^2.0.5 => 2.0.7
    gatsby-source-filesystem: ^2.0.2 => 2.0.12
    gatsby-transformer-remark: ^2.1.6 => 2.1.17
    gatsby-transformer-sharp: ^2.1.3 => 2.1.9
  npmGlobalPackages:
    gatsby-cli: 2.4.5
```

### PR DESCRIPTION
**Pull request by DSchau**
## Description

This PR upgrades our root Jest dependency to Jest 24 (released today!), and also does the following:

- Tweaks the docs/guides for unit testing for Jest 24
- Uses the default starter instead of blog starter for testing examples (we should make this a convention going forward, it just makes things easier)
- Tweaks the using-jest example for Jest 24 support

## Related Issues

Fixes #11267

### CODE REVIEW

File: docs/docs/testing-components-with-graphql.md [hunk 1]
```diff
@@ -30,39 +30,46 @@ done.
...
+import { StaticQuery } from "gatsby"
+import Index from "../pages/index"
 
-describe("BlogIndex", () =>
-  it("renders correctly", () => {
-    const location = {
-      pathname: "",
-    }
+beforeEach(() => {
+  StaticQuery.mockImplementationOnce(({ render }) =>
```
Comment (DSchau) [inline docs/docs/testing-components-with-graphql.md]: We should think about how we can improve this 🤔 It's very not beginner friendly.

File: package.json [hunk 2]
```diff
@@ -21,7 +21,7 @@
     "fs-extra": "^7.0.0",
     "glob": "^7.1.1",
     "husky": "1.1.1",
-    "jest": "^23.5.0",
+    "jest": "^24.0.0",
     "jest-cli": "^23.5.0",
```
Comment (stefanprobst) [inline package.json]: Should also update `babel-jest`, fixes #10913.

Comment (DSchau) [reply package.json]: good call; thanks!

Comment (DSchau) [reply package.json]: Yeah - this seems to break the [Windows pipeline](https://dev.azure.com/gatsbyjs/gatsby/_build/results?buildId=5081) 😩 

> TypeError: Cannot read property 'cwd' of undefined

@stefanprobst is there a substantive reason we should make this change?

Comment (stefanprobst) [reply package.json]: I remember having a problem with the version mismatch between `babel-jest@22` and `jest@23` in current master at some point, that's where the PR came from. Does updating `jest-cli` help?

File: packages/gatsby-plugin-canonical-urls/src/__tests__/__snapshots__/gatsby-ssr.js.snap [hunk 3]
```diff
@@ -16,7 +16,7 @@ exports[`Adds canonical link to head correctly creates a canonical link if siteU
   ],
   "results": Array [
     Object {
-      "isThrow": false,
+      "type": "return",
```
Comment (DSchau) [inline packages/gatsby-plugin-canonical-urls/src/__tests__/__snapshots__/gatsby-ssr.js.snap]: ... all of these are gonna cause some churn too 😩

File: package.json (hunk 2 above)
Comment (stefanprobst) [reply package.json]: And the snapshot changes seem to be because of https://github.com/facebook/jest/pull/6381

**Pull request merged**